File: clients/js/src/features/error_resolve_uri.ts
import { Input } from "../input";

import {
  PolywrapClientConfigBuilder,
  PolywrapClient,
  ExtendableUriResolver,
} from "@polywrap/client-js";
import { WasmPackage } from "@polywrap/wasm-js";

import path from "path";
import fs from "fs";

export async function runTestCase(input: unknown): Promise<void> {
  const inputObj = Input.expectObject<{
    resolver?: {
      directory: unknown;
      uri: unknown;
    };
    uri: unknown;
    expectedError: unknown;
  }>(input);

  const resolver = inputObj.resolver;
  const config = new PolywrapClientConfigBuilder();

  if (resolver) {
    const wrapDir = Input.expectRootDir(
      resolver.directory,
      path.join(__dirname, "../../../../")
    );

    const manifest = fs.readFileSync(path.join(wrapDir, "wrap.info"));
    const wasmModule = fs.readFileSync(path.join(wrapDir, "wrap.wasm"));
    const resolverUri = Input.expectUri(resolver.uri);
    const wrap = {
      package: WasmPackage.from(manifest, wasmModule),
      resolverUri,
    };
    config
      .setPackage(wrap.resolverUri.uri, wrap.package)
      .addInterfaceImplementation(
        ExtendableUriResolver.defaultExtInterfaceUris[0].uri,
        resolverUri.uri
      );
  }

  const uri = Input.expectUri(inputObj.uri);
  const client = new PolywrapClient(config.build());

  console.log(`Resolving URI ${uri.uri}`);

  const result = await client.invoke({
    uri: uri.uri,
    method: "",
  });

  const expectedError = Input.expectString(inputObj.expectedError);
  if (!result.ok) {
    console.log(result.error)
    if (result.error?.toString().toLowerCase().includes(expectedError.toLowerCase())) {
      console.log("Expected error received");
    } else {
      console.log(
        `Expected error "${expectedError}", but received "${result.error}"`
      );
    }
  }
}

File: clients/py/src/features/error_resolve_uri.py
from pathlib import Path
from typing import Any, Optional
from polywrap import (
    PolywrapClient,
    PolywrapClientConfigBuilder,
    SimpleFileReader,
    ExtendableUriResolver,
    WasmPackage,
)
from pydantic import BaseModel, Field, validator
from validators import UriStr, validate_root_directory

# Define the WrapDir structure as a Pydantic model
class WrapDir(BaseModel):
    directory: Path
    uri: UriStr

    @validator("directory")
    def valid_root_directory(cls, v: Any):
        return validate_root_directory(
            v, Path(__file__).parent.parent.parent.parent.parent
        )

# Define the input structure as a Pydantic model
class InputObj(BaseModel):
    resolver: Optional[WrapDir]
    uri: UriStr
    expected_error: str = Field(..., alias="expectedError")


def run_test_case(input: Any) -> None:
    # Parse the input to the defined model
    input_obj = InputObj.parse_obj(input)

    root_dir = Path(__file__).parent.parent.parent.parent.parent

    config_builder = PolywrapClientConfigBuilder()

    if input_obj.resolver:
        # Process resolver WrapDir
        resolver_directory = root_dir / input_obj.resolver.directory

        resolver_manifest_path = resolver_directory / "wrap.info"
        resolver_wasm_path = resolver_directory / "wrap.wasm"

        with open(resolver_manifest_path, "rb") as f:
            resolver_manifest = f.read()

        with open(resolver_wasm_path, "rb") as f:
            resolver_wasm_module = f.read()

        resolver_package = WasmPackage(
            wasm_module=resolver_wasm_module,
            file_reader=SimpleFileReader(),
            manifest=resolver_manifest,
        )

        config_builder.set_package(input_obj.resolver.uri, resolver_package)
        config_builder.add_interface_implementations(
            ExtendableUriResolver.DEFAULT_EXT_INTERFACE_URIS[0], [input_obj.resolver.uri]
        )

    config = config_builder.build()

    client = PolywrapClient(config)

    print(f"Resolving URI {input_obj.uri}")

    try:
        client.invoke(uri=input_obj.uri, method="", args=None)

    except Exception as error:
        if input_obj.expected_error in str(error):
            print("Expected error received")
        else:
            print(
                f"Expected error {input_obj.expected_error}, but received {str(error)}"
            )
